src/dataflow/core/type_inference.py

```diff
@@ -0,0 +1,330 @@
+import json
+from abc import ABC
+from dataclasses import dataclass, field, replace
+from typing import Dict, List, Optional, Sequence, Set, cast
+
+from dataflow.core.definition import Definition
+from dataflow.core.program import (
+    BuildStructOp,
+    CallLikeOp,
+    Expression,
+    Program,
+    TypeName,
+    ValueOp,
+    roots_and_reentrancies,
+)
+
+
+@dataclass(frozen=True)
+class TypeInferenceError(Exception):
+    msg: str
+
+
+class Type(ABC):
+    pass
+
+
+class TypeVariable(Type):
+    pass
+
+
+class NamedTypeVariable(TypeVariable):
+    """A named type variable like T. Note that named type variables have scope:
+    several different functions might use the name T but they are distinct type
+    variables. As such, reference identity for NamedTypeVariables matters."""
```

Comment:
:nitpick:
1. For multi-line docstrings, the first line should be a one-liner summary. See [PEP0257](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings).
2. Also it is more common to make the ending `"""` on its own line. See [PEP0008](https://www.python.org/dev/peps/pep-0008/#documentation-strings).

```suggestion
    """A named type variable like T. 
    
    Note that named type variables have scope:
    several different functions might use the name T but they are distinct type
    variables. As such, reference identity for NamedTypeVariables matters.
    """
```